### PR TITLE
Do not skip flaky test in CI

### PR DESCRIPTION
**Pull request by lukebakken**
Part of #77

### CODE REVIEW

**Comment by lukebakken** [comment]
cc @fho 

Well, this sure is interesting. With only 1 cpu in use the `TestRepeatedChannelExceptionWithPublishAndMaxProcsIssue46` test never fails, even without these changes. As soon as you add CPUs the likelihood of failures goes up. My guess is that this was never a problem until Go 1.15 because of how `GOMAXPROCS` used to work. On the GitHub runners there may only be 1 virtual CPU available so we wouldn't see this behavior without specifying `-cpu 1,2`.

Anyway, on my 16-core laptop, I tested this out with the following command:

```
go test -cpu 1,2,4,8,16 -failfast -race -v -tags integration -run TestRepeatedChannelExceptionWithPublishAndMaxProcsIssue46
```

This is a typical sequence of errors in the RabbitMQ logs:

```
2022-05-19 20:17:24.475000-07:00 [error] <0.29923.2> Channel error on connection <0.29739.2> (127.0.0.1:62619 -> 127.0.0.1:5672, vhost: '/', user: 'guest'), channel 22:
2022-05-19 20:17:24.475000-07:00 [error] <0.29923.2> operation basic.publish caused a channel exception not_found: no exchange 'not-existing-exchange' in vhost '/'
2022-05-19 20:17:24.476000-07:00 [error] <0.29739.2> Error on AMQP connection <0.29739.2> (127.0.0.1:62619 -> 127.0.0.1:5672, vhost: '/', user: 'guest', state: running), channel 22:
2022-05-19 20:17:24.476000-07:00 [error] <0.29739.2>  operation basic.publish caused a connection exception channel_error: "expected 'channel.open'"
2022-05-19 20:17:24.477000-07:00 [info] <0.29739.2> closing AMQP connection <0.29739.2> (127.0.0.1:62619 -> 127.0.0.1:5672, vhost: '/', user: 'guest')
2022-05-19 20:17:24.477000-07:00 [debug] <0.29942.2> Closing all channels from connection '127.0.0.1:62619 -> 127.0.0.1:5672' because it has been closed
```

The changes in this PR take into account the fact that you shouldn't use a channel after *any* channel exception (which was not the case prior to these changes). Notice the **connection exception** due to the fact that the Go client snuck in a `basic.publish` frame when the channel was closed.

So my guess at this time is that the increased concurrency available via `GOMAXPROCS > 1` makes it likely that "bad" data will make it to RabbitMQ even after a channel-level exception has occurred. Since the `TestRepeatedChannelExceptionWithPublishAndMaxProcsIssue46` test behaved abusively it made failures much more likely.

**Comment by lukebakken** [comment]
Cool, a 👍 from @fho is as good as any review. Merging once the tests pass. Have a great weekend.

**Comment by fho** [comment]
@lukebakken :sweat_smile: I actually only read your comment so far. 
I'll have a deeper look next week and let you know if I find something post-mortem.
have a nice weekend! :-)

**Comment by lukebakken** [comment]
Sounds good, thanks.

File: integration_test.go [hunk 1]
```diff
@@ -1540,25 +1539,46 @@ func TestDeadlockConsumerIssue48(t *testing.T) {
...
-			if err != nil {
-				t.Fatalf("expected error only on publish, got error on channel.open: %v", err)
+	t.Cleanup(func() {
+		if conn != nil {
+			conn.Close()
+		}
+	})
+
+	for i := 0; i < 100; i++ {
+		if conn == nil || conn.IsClosed() {
```
Comment (fho) [inline integration_test.go:1551]: Why is this changed to open a new connection if the connections got closed?
Before it was expected in the testcase that the connection does not break and is kept open

File: integration_test.go [hunk 2]
```diff
@@ -1540,25 +1539,46 @@ func TestDeadlockConsumerIssue48(t *testing.T) {
...
+			conn = integrationConnection(t, "issue46")
+			if conn == nil {
+				t.Fatal("conn is nil")
+			}
```
Comment (fho) [inline integration_test.go:1555]: How could this condition happen?
conn is never changed after it is opened, and integrationConnection() always returns a non-nil conn

File: integration_test.go [hunk 3]
```diff
@@ -1540,25 +1539,46 @@ func TestDeadlockConsumerIssue48(t *testing.T) {
...
+		}
+
+		ch, err := conn.Channel()
+		if err, ok := err.(Error); ok {
+			if err.Code != 504 {
+				t.Fatalf("expected channel only exception i: %d got: %+v", i, err)
 			}
+		}
```
Comment (fho) [inline integration_test.go:1563]: Why was this changed from not allowing any errors to allow all errors except 504?
Doesn't this change the behavior of the test and prevents that it catches the original bug?

File: channel.go [hunk 4]
```diff
@@ -230,6 +230,11 @@ func (ch *Channel) sendOpen(msg message) (err error) {
...
 		}
 
+		// If the channel is closed, use Channel.sendClosed()
+		if atomic.LoadInt32(&ch.closed) == 1 {
+			return ch.sendClosed(msg)
+		}
+
 		if err = ch.connection.send(&methodFrame{
 			ChannelId: ch.id,
 			Method:    content,
```
Comment (fho) [inline channel.go:240]: Isn't this vulnerable to a race condition?
If the channel gets closed after the `ch.closed` check send will be called on the closed channel

**Review by fho — COMMENTED** [review]
The PR changes the `TestRepeatedChannelExceptionWithPublishAndMaxProcsIssue46` testcase in a way that it does not fail on the same condition then before.

The actual bug, that a connection becomes unusable after a message is published on a channel for a non-existing exchange still happens from now and then.
I can reproduce the issue with the previous version of the `TestRepeatedChannelExceptionWithPublishAndMaxProcsIssue46` testcase via `go test -count=100 -cpu 24 -race -v -tags integration -run=TestRepeatedChannelExceptionWithPublishAndMaxProcsIssue46 ./...` on top of ffeb46d

File: integration_test.go (hunk 3 above)
Comment (lukebakken) [reply integration_test.go:1563]: In the case of a 504 you need a new connection.

File: channel.go (hunk 4 above)
Comment (lukebakken) [reply channel.go:240]: It's less racy than before!

**Comment by lukebakken** [comment]
> The actual bug, that a connection becomes unusable after a message is published on a channel for a non-existing exchange

Publishing to an exchange that doesn't exist causes a channel-level exception from RabbitMQ. After such an exception, you **can't** continue to use that channel. The test used to continue publishing to the same channel which violates the protocol.

In addition, the test might also sneak a `basic.publish` to RabbitMQ after RabbitMQ closes the channel (due to the channel exception), in which case you cause a **connection**-level exception, which is why the test must also test for this.

So, the original test was flawed but it worked because the tests always used to run with `GOMAXPROCS=1`.

Really this test exposes why a RabbitMQ publisher _must_ use publisher confirms, or at a bare minimum, the `mandatory` flag, and to also not use a channel after a channel-level exception.

We can continue this in a new issue if you'd like.

File: channel.go (hunk 4 above)
Comment (lukebakken) [reply channel.go:240]: Good point, let me re-visit this.

Comment (lukebakken) [reply channel.go:240]: I think my changes are correct, based on this comment -
https://github.com/rabbitmq/amqp091-go/blob/main/channel.go#L108-L110

Basically, by checking `ch.closed` at the last possible time in these two places I've narrowed the window where we could be sending the wrong data to RabbitMQ.

**Comment by lukebakken** [comment]
@fho I just woke up in the middle of the night and thought again about the test and realized there is indeed a bug somewhere in the library. Then of course you've already commented to say as much.

I'll open a new issue in the morning. Thanks!

**Comment by lukebakken** [comment]
https://github.com/rabbitmq/amqp091-go/pull/87
